Firestore/Swift/Source/PropertyWrapper/FirestoreQuery.swift

```diff
@@ -0,0 +1,176 @@
+    .isGreaterThanOrEqualTo(field, value)
+  }
+
+  public static func order(by field: String, descending value: Bool = false) -> QueryPredicate {
+    .orderBy(field, value)
+  }
+
+  public static func limit(to field: Int) -> QueryPredicate {
+    .limitTo(field)
+  }
+}
+
+@available(iOS 13.0, *)
+internal class FirestoreQueryObservable<T: Decodable>: ObservableObject {
+  @Published var items: [T] = []
+
+  private let firestore = Firestore.firestore()
+  private var listener: ListenerRegistration? = nil
+
+  init(collectionPath: String, predicates: [QueryPredicate]) {
+    setupListener(from: collectionPath, withPredicates: predicates)
+  }
+
+  deinit {
+    removeListener()
+  }
+
+  private func setupListener(from collectionPath: String,
+                             withPredicates predicates: [QueryPredicate]) {
+    var query: Query = firestore.collection(collectionPath)
+
+    for predicate in predicates {
+      switch predicate {
+      case let .isEqualTo(field, value):
+        query = query.whereField(field, isEqualTo: value)
+      case let .isIn(field, values):
+        query = query.whereField(field, in: values)
+      case let .isNotIn(field, values):
+        query = query.whereField(field, notIn: values)
+      case let .arrayContains(field, value):
+        query = query.whereField(field, arrayContains: value)
+      case let .arrayContainsAny(field, values):
+        query = query.whereField(field, arrayContainsAny: values)
+      case let .isLessThan(field, value):
+        query = query.whereField(field, isLessThan: value)
+      case let .isGreaterThan(field, value):
+        query = query.whereField(field, isGreaterThan: value)
+      case let .isLessThanOrEqualTo(field, value):
+        query = query.whereField(field, isLessThanOrEqualTo: value)
+      case let .isGreaterThanOrEqualTo(field, value):
+        query = query.whereField(field, isGreaterThanOrEqualTo: value)
+      case let .orderBy(field, value):
+        query = query.order(by: field, descending: value)
+      case let .limitTo(field):
+        query = query.limit(to: field)
+      }
+    }
+
+    listener = query.addSnapshotListener { [weak self] snapshot, error in
+      if let error = error {
+        print(error)
+        self?.items = []
+        return
+      }
+
+      guard let snapshot = snapshot else {
+        print("FirestoreQuery: Registering the SnapshotListener returned a bad snapshot.")
+        self?.items = []
+        return
+      }
+
+      self?.items = snapshot.documents.compactMap { document in
+        try? document.data(as: T.self)
+      }
+    }
+  }
+
+  private func removeListener() {
+    listener?.remove()
+  }
+}
+
+@available(iOS 14.0, *)
+@propertyWrapper
+public struct FirestoreQuery<T: Decodable>: DynamicProperty {
+  @StateObject private var firestoreQueryObservable: FirestoreQueryObservable<T>
+
+  public var wrappedValue: [T] {
+    get {
+        firestoreQueryObservable.items
+    }
+  }
+
```

Comment:
In my opinion it would be beneficial to be able to override 'firestore` as well as the 'decoder' in use. Something like:

```
    public var firestore: Firestore {
        get { store.firestore }
        set { store.firestore = newValue }
    }

    public var decoder: Firestore.Decoder {
        get { store.decoder }
        set { store.decoder = newValue }
    }

```

and then in the `FirestoreQueryObservable`:
```
    fileprivate var firestore: Firestore = Firestore.firestore() {
        didSet {
            resubscribe()
        }
    }
    fileprivate var decoder: Firestore.Decoder = Firestore.Decoder() {
        didSet {
            resubscribe()
        }
    }

    private func resubscribe() {
        removeListener()
        self.items = []
        setupListener(from: collectionPath, withPredicates: predicates)
    }

```
This of course requires you to store the collectionPath and predicates on the `FirestoreQueryObservable`.
And the `removeListener` should then also nil out the listener...

What do you think?

Reply:
Hi Morten, more configuration options add more flexibility, but also more liability. Can you elaborate why you'd want to be able to configure the Firestore instance and the decoder? This will help us decide whether to add this to this PR or add it later.

Reply:
Main reason is just parity between this and other versions of the same API (the API meaning: fetching data from a Firestore collection somehow).
If `Firestore.Decoder` implemented different 'strategies' like `JSONDecoder` and also `Database.Decoder`, then there would naturally be an even bigger reason.
The complexity could be alleviated a bit by only supporting overrides through the initializer - although property wrapper initializers are 'hard' to inject values into...

Reply:
> It would be great to get this PR merged.
> 
> Besides tests, what else needs to be done, what are the next steps, and are there any blockers?

@peterfriese What do you think here?